### PR TITLE
 Implemented return in call stub

### PR DESCRIPTION
**Pull request by tsarn**
I didn't test it, might be broken

### CODE REVIEW

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 1]
```diff
@@ -295,29 +294,23 @@ class StubGenerator: public StubCodeGenerator {
...
+               r_entryframe_fp, _top_ijava_frame_abi_neg(result_address));
+      __ ld(r_arg_result_type,
+               r_entryframe_fp, _top_ijava_frame_abi_neg(result_type));
+
+      __ ld(R1_RA, r_entryframe_fp, _top_ijava_frame_abi_neg(ra));
+      __ ld(r_temp, r_entryframe_fp, _top_ijava_frame_abi_neg(fp));
+
+      // pop frame and restore non-volatiles, SP and FP
+      __ mv(R2_SP, r_entryframe_fp);
+      __ mv(R8_FP, r_temp);
```
Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: Can we restore directly to `R8` (from SP-16)?

**Review by AntonKozlov — COMMENTED** [review]
I suggest to verify the current code and then shuffle it for a bit to make more straightforward. 

Interesting, that GCC (I suppose it should be idiomatic for such simple cases) do not use `fp` for own needs and optimize it away (on `-O3`) unless requested (-fno-omit-frame-pointer -O3). 

(1) Restoring everything before `sp`/`fp` would make intention more clear. (2) Or, in opposite, restore sp/fp first and the everything else (like pushed). 
However, it's better to have a stack frame as long as possible (i.e. for accidental profiler not to catch inconsistent `pc`/`fp` state). So I vote for (1), as generated by GCC 
```
extern int f();

int main(void) {
        volatile char t[64];
        f(t);
        return 1;
}

riscv64-linux-gnu-gcc -fno-omit-frame-pointer -O3 -c test.c -o test.o
0000000000000000 <main>:
   0:	715d                	addi	sp,sp,-80
   2:	e0a2                	sd	s0,64(sp)
   4:	e486                	sd	ra,72(sp)
   6:	0880                	addi	s0,sp,80
   8:	fb040513          	addi	a0,s0,-80
   c:	00000097          	auipc	ra,0x0
  10:	000080e7          	jalr	ra # c <main+0xc>
  14:	60a6                	ld	ra,72(sp)
  16:	6406                	ld	s0,64(sp)
  18:	4505                	li	a0,1
  1a:	6161                	addi	sp,sp,80
  1c:	8082                	ret

File: src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp [hunk 2]
```diff
@@ -772,35 +772,53 @@ void InterpreterMacroAssembler::merge_frames(Register Rsender_sp, Register retur
...
 
   bind(notByte);
-  cmpwi_PPC(CCR0, ret_type, T_CHAR);
-  bne_PPC(CCR0, notChar);
-  andi(result, result, 0xffff);
-  b_PPC(done);
+  addi(scratch, R0, T_CHAR);
+  bne(scratch, ret_type, notChar);
+  addi(scratch, R0, 0xfff);
+  lui(scratch, 0xf);
```
Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp]: I would propose use `load_const` or similar, becasue it's potentially more efficient and it should have fewer bugs
* it should be `lui` + `addi`, as lui clears low bits
* 0xfff will be sign extended, so you need to load a (value+1) with lui

File: src/hotspot/cpu/riscv/templateTable_riscv_64.cpp [hunk 3]
```diff
@@ -2177,13 +2178,13 @@ void TemplateTable::_return(TosState state) {
 
     // Check if the method has the FINALIZER flag set and call into the VM to finalize in this case.
     assert(state == vtos, "only valid state");
-    __ ld_PPC(R25_tos, 0, R26_locals);
+    __ ld(R25_tos, R26_locals, 9);
```
Comment (Geckostya) [inline src/hotspot/cpu/riscv/templateTable_riscv_64.cpp]: why 9?

File: src/hotspot/cpu/riscv/macroAssembler_riscv.cpp [hunk 4]
```diff
@@ -3017,21 +3016,17 @@ void MacroAssembler::decode_klass_not_null(Register dst, Register src) {
...
 void MacroAssembler::load_klass(Register dst, Register src) {
-  if (UseCompressedClassPointers) {
-    lwz_PPC(dst, oopDesc::klass_offset_in_bytes(), src);
-    // Attention: no null check here!
-    decode_klass_not_null(dst, dst);
-  } else {
-    ld_PPC(dst, oopDesc::klass_offset_in_bytes(), src);
-  }
+  // TODO_RISCV: add support for compressed class pointers
+  assert(!UseCompressedClassPointers, "RISCV port: compressed class pointers not supported");
```
Comment (Geckostya) [inline src/hotspot/cpu/riscv/macroAssembler_riscv.cpp]: It is better to save previous `if` in this code to implement this later

**Comment by Geckostya** [comment]
I've tested it with void and integer values, it works.
Please, review it. @TsarN @AntonKozlov @vasyoid @DmKazakov  @sreznick

File: src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp [hunk 5]
```diff
@@ -743,64 +743,53 @@ void InterpreterMacroAssembler::unlock_if_synchronized_method(TosState state,
...
+  bne(scratch, ret_type, notByte);
+  // sign-extend lower 8 bits
+  li(scratch, 0x80);
+  bge(result, scratch, byteOne);
+  andi(result, result, 0xff);
+  j(done);
+  bind(byteOne);
+  li(scratch, 0xffffff00);
+  orr(result, result, scratch);
+  j(done);
```
Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp]: cpu/x86/interp_masm_x86.cpp, 
```
  bind(notBool);
  cmpl(rcx, T_BYTE);
  jcc(Assembler::notEqual, notByte);
  LP64_ONLY(movsbl(result, result);)
  NOT_LP64(shll(result, 24);)      // truncate upper 24 bits
  NOT_LP64(sarl(result, 24);)      // and sign-extend byte
  jmp(done);
```
`shll` -- logical shift left, `srai` -- arithmetic shift right. 
Can we implement this?

File: src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp [hunk 6]
```diff
@@ -743,64 +743,53 @@ void InterpreterMacroAssembler::unlock_if_synchronized_method(TosState state,
...
-  extsh_PPC(result, result);
+  // sign-extend lower 16 bits
+  li(scratch, 0x8000);
+  bge(result, scratch, shortOne);
+  li(scratch, 0xffff);
+  andr(result, result, scratch);
+  j(done);
+  bind(shortOne);
+  li(scratch, 0xffff0000);
+  orr(result, result, scratch);
```
Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp]: can be simplified as well

Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp]: can be simplified as well

File: src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp [hunk 7]
```diff
@@ -2364,7 +2355,7 @@ void InterpreterMacroAssembler::verify_oop_or_return_address(Register reg, Regis
   mr_PPC(R3_ARG1_PPC, R27_method);
   call_c(Rtmp); // call C
 
-  pop_frame();
+  pop_C_frame();
   restore_LR_CR(Rtmp);
```
Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp]: Shouldn't be these two lines `pop_C_frame(true/*restore_ra*/)`?

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp [hunk 8]
```diff
@@ -286,29 +285,20 @@ class StubGenerator: public StubCodeGenerator {
...
+      assert_different_registers(R10_RET1, r_arg_result_addr, r_arg_result_type);
+
+      __ ld(r_arg_result_addr, R8_FP, _entry_frame_locals(result_address));
+      __ ld(r_arg_result_type, R8_FP, _entry_frame_locals(result_type));
+
+
+      // pop frame and restore non-volatiles, SP and FP
+      __ mv(R2_SP, R8_FP);
+      __ ld(R1_RA, R8_FP, _abi(ra));
+      __ ld(R8_FP, R8_FP, _abi(fp));
```
Comment (AntonKozlov) [inline src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: `pop_C_frame(true)`?

File: src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp (hunk 7 above)
Comment (Geckostya) [reply src/hotspot/cpu/riscv/interp_masm_riscv_64.cpp]: Yeah, you're right

File: src/hotspot/cpu/riscv/stubGenerator_riscv.cpp (hunk 8 above)
Comment (Geckostya) [reply src/hotspot/cpu/riscv/stubGenerator_riscv.cpp]: It is an entry frame, not C. I don't want to mislead the reader. It is better to change in this way:
```
      __ mv(R2_SP, R8_FP);
      __ restore_abi_frame(R8_FP, 0);
```

**Review by Geckostya — APPROVED** [review]

**Pull request merged**